proposals/0351-regex-builder.md

```diff
@@ -428,16 +441,204 @@ extension RegexComponentBuilder {
+// Equivalent: '(CREDIT|DEBIT)'
+Capture {
+  ChoiceOf {
+    "CREDIT"
+    "DEBIT"
+  }
+} // `.RegexOutput == (Substring, Substring)`
+```
+
+To transform the captured substring, specify a `transform:` closure.
+
+```
+// This example is similar to the one above, however in this example we
+// transform the result of the capture into:
+// "Transaction Kind: CREDIT" or "Transaction Kind: DEBIT"
+Capture {
+  ChoiceOf {
+    "CREDIT"
+    "DEBIT"
+  }
+} transform: {
+  "Transaction Kind: \($0)"
+} // `.RegexOutput == (Substring, String)`
+```
+
+The transform closure can throw. When a transform closure throws during matching, the matching will abort and the error will be propagated directly to the top-level matching API that's being called, e.g. `Regex.wholeMatch(in:)` and `Regex.prefixMatch(in:)`.
+
+An alternative version of capture is called `TryCapture`, which works in cases where you want to transform the capture, but the transformation may return nil. When a nil is returned, the regex engine backtracks and tries an alternative. `TryCapture` is most useful when you transform a captured substring by calling a failable initializer.
+
+```swift
+enum TransactionKind: String {
+  case credit = "CREDIT"
+  case debit = "DEBIT"
+}
+
+TryCapture {
+  ChoiceOf {
+    "CREDIT"
+    "DEBIT"
+  }
+} transform: {
+  // This initializer may return nil which is why we used TryCapture.
+  TransactionKind(rawValue: String($0))
+}
+```
+
+Reference is a feature that can be used to achieve named captures and named backreferences from textual regexes. Simply state what type the reference will hold on to and you can use it later once you've matched a string to get back a specific capture. Note the type you pass to reference will be whatever the result of a capture's transform is. A capture with no transform always has a reference type of `Substring`.
+
+```swift
+let kind = Reference(Substring.self)
+
+let regex = Capture(as: kind) {
+  ChoiceOf {
+    "CREDIT"
+    "DEBIT"
+  }
+}
+
+let input = "CREDIT"
+if let result = regex.firstMatch(in: input) {
+  print(result[kind]) // CREDIT
+}
+```
+
+When capturing some regex with a reference specified, the reference will refer to the most recently captured content. The reference itself can also be used as a regex to match the most recently captured content, or as a name to look up the result of matching.
+
+```swift
+let a = Reference(Substring.self)
+let b = Reference(Substring.self)
+let regex = Regex {
+  Capture("abc", as: a)
+  Capture("def", as: b)
+  a
+  Capture(b)
+}
+
+if let result = regex.firstMatch(in: input) {
+  print(result[a]) // => "abc"
+  print(result[b]) // => "def"
+}
+```
+
+A regex is considered invalid when it contains a use of reference without it ever being captured in the regex. When this occurs in the regex builder DSL, a runtime error will be reported. Similarly, the use of a reference in a `Regex.Match.subscript(_:)` must have been captured in the regex that produced the match.
```

Comment:
I think it's important to very clearly differentiate the `throw` of a runtime error during the match call and the trap that's generated in the subscript. As written, the reader has to read a little between the lines here.

Reply:
Wait, is that what happens? I thought it would just give `nil`.

Reply:
Can we clarify what's going on? I would find it incredibly developer hostile for this to trap if an alternation branch wasn't taken. Perhaps this should say "trap if the reference was not used in the regex" rather than captured.

Reply:
> Can we also be clear that this means that a reference that is captured only in an optional portion of a `Regex` can trap when accessed? Do we have API for knowing whether or not it is safe to access?

This behavior is unfortunate and could definitely use some more discussion.  I see some options:
1. Make the subscript return optional, but it could make things harder to use than numbered captures for non-optional captures.
    ```swift
    let ref = Reference(Substring.self)
    let regex = Regex {
          Capture(..., as: ref)
    }
    let match = try regex.fisrtMatch(of: ...)!
    match.1 // Substring
    match[ref] // Substring?
    ```
2. Provide an API to check whether a match result contains a reference. The user would be responsible for checking when the reference capture is in an optional branch of the regex. This can keep non-optional captures simple.
    ```swift
    // Non-optional trapping version
    subscript<Capture>(_ reference: Reference<Capture>) -> Capture { get }
    // Checking version
    subscript<Capture>(ifExists reference: Reference<Capture>) -> Capture? { get }
    // Predicate
    func contains<Capture>(_ reference: Reference<Capture>) -> Bool
    ```

    ```swift
    let ref = Reference(Substring.self)
    let regex = Regex {
          ZeroOrMore(Capture(..., as: ref))
    }
    let match = try regex.fisrtMatch(of: ...)!
    match.1 // Substring?
    match[ref] // Substring or trap
    match[ifExists: ref] // Substring?
    match.contains(ref) // Bool
    ```

Reply:
Went with option 1.